### DISCUSSION TITLE
Should I also stop reading characteristic when disconnecting?

### DISCUSSION THREAD

**Discussion by alessandrofrancesconi** [dotnet-bluetooth-le/dotnet-bluetooth-le#796] [Q&A]
I'm using Plugin.BLE 3.0.0

When connecting to a known device, I also start listening to a particular characteristic. Simplified code:

```
// do connection
this.Ble.Adapter.ConnectToKnownDeviceAsync(<guid...>);
<...>
// get services and start listen to data from a given Characteristic:
var readChar = await service.GetCharacteristicAsync(<guid...>);
readChar.ValueUpdated += ReadCharacteristic_ValueUpdated;
_ = readChar.StartUpdatesAsync();
```

At a certain point, I will disconnect from the device with:

`await this.Ble.Adapter.DisconnectDeviceAsync(<...>);`

Should I also call `readChar.StopUpdatesAsync()` here? Or is it handled internally by the library?
If I must stop it, should I do it before or after disconnection?

Thanks for your help

**Comment by cckring** [comment]
@alessandrofrancesconi have you found an answer?

**Reply by alessandrofrancesconi** [reply]
The only thing I've seen is an "hint" inside this issue: https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/issues/226

> Basically, it turned out that it's necessary to call ICharacteristic.StopUpdatesAsync() for every ICharacteristic.StartUpdatesAsync() before calling IAdapter.DisconnectDeviceAsync(). I suggest making the de-registering of those callbacks internal to IAdapter.DisconnectDeviceAsync(), rather than having the application call ICharacteristic.StopUpdatesAsync(). At the very least, this should be added to the documentation.

**Reply by axa88** [reply]
If _not_ calling `ICharacteristic.StopUpdatesAsync()` for every `ICharacteristic.StartUpdatesAsync()` causes a failure to deregister callbacks in the OS, then that is a separate issue to whether or not to request these notifications are stopped. [see](https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/discussions/796#discussioncomment-9332045)

**Comment by axa88** [comment]
Ok version 2 of my answer...

>Should I also call readChar.StopUpdatesAsync() here? Or is it handled internally by the library?

first understand what your doing by calling `StopUpdatesAsync` is telling the peripheral to take this device out of the list of devices that have asked _to be_ notified of a data change.

I dug into my ble device firmware and noticed that the notification sub is volatile unless the device is bonded.
Therefore what you should do entirely depends on your intent and and bond capabilities.
And the library isn't (shouldn't) be doing anything like this for you.

So expect that if not bonded it doesn't matter what you choose to do upon disconnecting as notifications will always be disabled.
And if you are bonded, expect that if you dont call stop updates then youll get them again when reconnecting.


> If I must stop it, should I do it before or after disconnection?

You can only ask yourself if you must stop, and if you decide you must stop then you must do it before disconnecting (or after reconnecting in the case of a bonded device)

Also [see](https://community.silabs.com/s/question/0D51M00007xeQHUSA2/disable-able-notification-after-ble-disconnect?language=en_US)
